Core/include/Acts/Navigation/MultiWireLayerUpdators.hpp

```diff
@@ -0,0 +1,199 @@
+    float rangeY1 = min_max_multilayer[1].first - radius,
+          rangeY2 = min_max_multilayer[1].second + radius;
+
+    std::size_t nBinsX =
+                    static_cast<std::size_t>((rangeX2 - rangeX1) / binWidthX),
+                nBinsY =
+                    static_cast<std::size_t>((rangeY2 - rangeY1) / binWidthY);
+
+    detail::IndexedSurfacesGenerator<decltype(gridSurfaces)> irSurfaces_grid{
+        gridSurfaces, {}, {binZ, binY}, {1u, 0u}};
+
+    detail::GridAxisGenerators::EqBoundEqBound aGenerator_grid{
+        {rangeX1, rangeX2}, nBinsX, {rangeY1, rangeY2}, nBinsY};
+
+    detail::CenterReferenceGenerator rGenerator;
+
+    auto indexedSurfaces = irSurfaces_grid(gctx, aGenerator_grid, rGenerator);
+
+    using GridType =
+        decltype(aGenerator_grid)::grid_type<std::vector<std::size_t>>;
+    using DelegateType = IndexedSurfacesAllPortalsImpl<GridType>;
+
+    const auto* instance_multilayer = indexedSurfaces.instance();
+    auto castedDelegate_grid =
+        dynamic_cast<const DelegateType*>(instance_multilayer);
+
+    const auto& chainedUpdators_grid = castedDelegate_grid->updators;
+    const auto& indexedSurfaces_multilayer =
+        std::get<IndexedSurfacesImpl<GridType>>(chainedUpdators_grid);
+
+    return indexedSurfaces_multilayer;
+  }
+
+  inline void update(const Acts::GeometryContext& gctx,
+                     Acts::Experimental::NavigationState& nState) const {
+    if (nState.currentVolume == nullptr) {
+      throw std::runtime_error(
+          "MultiWireLayerImpl: no detector volume set to navigation state.");
+    }
+
+    // A volume switch has happened, update list of portals & surfaces
+
+    if (nState.surfaceCandidates.empty()) {
+      // Fill internal portals if existing
+      for (const auto v : nState.currentVolume->volumes()) {
+        const auto& iPortals = v->portals();
+        Acts::Experimental::PortalsFiller::fill(nState, iPortals);
+      }
+
+      // Fill with the indexed surfaces and the portals of the volume
+
+      const auto m_indexedSurfaces = getIndexedSurfacesFromVolume(nState);
+
+      auto tposition = nState.position;
+      auto tdirection = nState.direction;
+
+      const auto& grid = m_indexedSurfaces.grid;
+      const auto& extractor = m_indexedSurfaces.extractor;
+      auto nBins = grid.numLocalBins();
+      auto binWidth = grid.binWidth();
+      auto minPos = grid.minPosition();
+      auto maxPos = grid.maxPosition();
+
+      Vector3 origin = Vector3(0, minPos[1], 0);
+      Translation3 trans(origin);
+      Transform3 transform(trans);
+
+      auto dy = binWidth[1];
+      auto phi = Acts::VectorHelpers::phi(tdirection);
+
+      BinUtility binUtility2D(nBins[0], minPos[0], maxPos[0], open, binX,
+                              transform);
+      binUtility2D +=
+          BinUtility(nBins[1], minPos[1], maxPos[1], open, binY, transform);
+
+      for (std::size_t i = 0; i < nBins[1]; i++) {
```

Comment:
This seems expensive to do on every update, right?

Reply:
> This seems expensive to do on every update, right?

Oh sorry, you mean the for loop? You are right, but I just needed to go through every layer of the grid